src/v1/Models/Item.php

```diff
@@ -46,6 +48,29 @@ class Item extends Model
     // 'getItems'
   ];
 
+  public static function boot()
+  {
+    parent::boot();
+    static::creating(function ($model)
+    {
+      /**
+       * Manage id_bytype field
+       * Get max id in DB for this type and increase 1 to get new id_bytype
+       */
+      $subquery = DB::table((new self())->getTable() . ' as item_alias')
+        ->selectRaw('coalesce(max(id_bytype), 0) + 1 as id_bytype')
+        ->where('type_id', $model->type_id);
+      $model->id_bytype = DB::raw("(" . vsprintf(
+        str_replace(
+          ['?'],
+          ['\'%s\''],
+          $subquery->toSql()
+        ),
+        $subquery->getBindings()
+      ) . ")");
```

Comment:
I don't understand why you need to do that instead of simply calling `get()` (disclaimer: I never worked with Laravel query builder).

If `get()` is not enough, at least I hope there's a better solution. Maybe something inspired by https://fideloper.com/laravel-raw-queries:

```php
$raw_query = DB::raw('SELECT coalesce(max(id_bytype), 0) + 1 AS id_bytype FROM items WHERE type_id = :type_id');
// note that I think it will not return the id directly, but probably in an array?
$model->id_bytype = DB::select($raw_expression, [
  ':type_id' => $model->type_id,
]);
```

Reply:
If you do the `get()` you will have the max id and then do the insert. So yes it works, but only if you not have same process running in same time (for example 2 users create a ticket at same time). Doing the subquery directly in the insert SQL prevent this problem. If have a problem, it return an error and that's why I catch the error and try pass again the query 4 times.

I have tested with 30 users in same times creating same type of item and no problems with this piece of code.

The code you suggest seems have too this problem on concurrent inserts.

Reply:
Do you mean that the actual value of `id_bytype` IS the SQL query and so it is executed during the insert?

I think generated columns could work for that, but it is only for MariaDB, MySQL and PostgreSQL 12+ (see https://laravel.com/docs/9.x/migrations#column-modifiers, `storedAs`).

If it doesn't work, at least we can try to not pass the dynamic `$model->type_id` with: `whereRaw('items.type_id = item_alias.type_id')`

Another option would be to do an update of the record during the `created` event:

```sql
UPDATE items
SET id_bytype=(the subquery)
WHERE id = :id
```

But we could end with something like that:

```
item id=1 id_bytype=2 type=1
item id=2 id_bytype=1 type=1
```

I.e. ids are not in the same order, but it seems it works for me

Reply:
Yes the select is executed in same time than insert, it's a subquery.

I will check the storedAs if it's possible.

the `whereRaw` is not a real solution like you said to prevent SQL injection. That's why I use `where`

do the subquery in the insert or do the insert and then do the subquery in update, I prefer doing it directly in the insert, it will prevent more problems :p

Reply:
Not possible use storedAs (generated columns) because we need a subquery and not possible with that (it's only for operations on fields of the row like concat 2 columns, or do an operation to have a percent...).

Seems not have better solution

Reply:
> the whereRaw is not a real solution like you said to prevent SQL injection. That's why I use where

There is no SQL injection possible in `whereRaw('items.type_id = item_alias.type_id')` (but maybe you can do `where('items.type_id', 'item_alias.type_id')`?)

Reply:
Yes works and it's better ;)
I pushed the modification ^_^

Reply:
the whereraw works here because not use a variable inside it, so can't have an injection. I tested with the `where()` and seems not possible to define a columns name inside

Reply:
Okay, it looks better :+1: (even if setting the value of `id_bytype` to a SQL expression still gives me cold sweats :sweat_smile:) No need to explicit `item_alias`?

Reply:
> the whereraw works here because not use a variable inside it, so can't have an injection

`whereRaw` also accepts an array of bindings, see https://laravel.com/docs/9.x/queries#whereraw-orwhereraw

Reply:
Finally, I use a raw query to be more simple and use the type_id stored in `$model`. I pass this variable into an `intval()` to be sure it's an integer and prevent SQL injection.